File: core/src/main/java/io/github/bindglam/neko/content/item/ItemImpl.java
package io.github.bindglam.neko.content.item;

import io.github.bindglam.neko.Neko;
import io.github.bindglam.neko.content.AbstractContent;
import io.github.bindglam.neko.content.feature.FeatureBuilder;
import io.github.bindglam.neko.content.item.properties.ItemProperties;
import io.github.bindglam.neko.platform.PlatformAdapter;
import io.github.bindglam.neko.platform.PlatformItemStack;
import io.github.bindglam.neko.platform.PlatformPersistentDataContainer;
import lombok.Getter;
import lombok.experimental.Accessors;
import net.kyori.adventure.key.Key;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

public final class ItemImpl extends AbstractContent implements Item {
    @Getter @Accessors(fluent = true)
    private final ItemProperties properties;
    @Getter @Accessors(fluent = true)
    private final ImmutableItemStack itemStack;

    public ItemImpl(@NotNull Key key,
                    @NotNull ItemProperties properties,
                    @NotNull List<FeatureBuilder> features) {
        super(key, features);
        this.properties = properties;
        this.itemStack = ImmutableItemStack.of(ItemBuilder.create(this));
    }

    @Override
    public boolean isSimilar(@NotNull PlatformItemStack itemStack) {
        boolean[] result = new boolean[] { false };
        itemStack.persistentDataContainer(persistentDataContainer ->
                result[0] = Objects.equals(key().asString(), persistentDataContainer.get(ItemBuilder.NEKO_ITEM_KEY, PlatformAdapter.adapter().persistentDataType(String.class).orElseThrow())));
        return result[0];
    }

    @Override
    public @NotNull String translationKey() {
        return "item." + key().namespace() + "." + key().value();
    }

    @Override
    public @NotNull Item asItem() {
        return this;
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/content/item/ItemImpl.kt
package io.github.bindglam.neko.content.item

import io.github.bindglam.neko.content.item.properties.ItemProperties
import net.kyori.adventure.key.Key
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType

class ItemImpl(private val key: Key, private val properties: ItemProperties) : Item {
    private val itemStack = ImmutableItemStack.of(ItemBuilder.create(this))

    override fun isSimilar(itemStack: ItemStack): Boolean {
        if(!itemStack.hasItemMeta()) return false
        return itemStack.itemMeta.persistentDataContainer.get(ItemBuilder.NEKO_ITEM_KEY, PersistentDataType.STRING) == key.asString()
    }

    override fun properties() = properties
    override fun itemStack() = itemStack
    override fun key() = key
    override fun translationKey() = "item.${key.namespace()}.${key.value()}"
    override fun asItem() = this
}